pipeline/transform.py

```diff
@@ -0,0 +1,193 @@
+"""Transform script"""
+from datetime import datetime
+import re
+import pandas as pd
+
+LATITUDE_INDEX = 0
+LONGITUDE_INDEX = 1
+ORIGIN_TOWN_INDEX = 2
+ORIGIN_COUNTRY_CODE_INDEX = 3
+ORIGIN_REGION_INDEX = 4
+
+
+def get_botanist_detail(reading: dict, botanist_detail: str) -> str | None:
+    """
+    Fetches specific detail about the botanists
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    if not isinstance(botanist_detail, str):
+        raise TypeError('botanist_detail must be of type str')
+    if 'botanist' in reading:
+        return reading['botanist'].get(botanist_detail)
+    return None
+
+
+def get_origin_detail(reading: dict, index: int) -> str | None:
+    """
+    Fetches specific detail about the plant's origin
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    if not isinstance(index, int):
+        raise TypeError('index must be of type int')
+    try:
+        return reading['origin_location'][index]
+    except KeyError:
+        return None
+
+
+def get_scientific_name(reading: dict) -> str | None:
+    """
+    Fetches scientific name of the plant
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    try:
+        return reading['scientific_name'][0]
+    except KeyError:
+        return None
+
+
+def get_origin_region(reading: dict) -> str | None:
+    """
+    Fetches the region of where the plant originated from
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    try:
+        return reading['origin_location'][ORIGIN_REGION_INDEX].split('/')[0]
+    except KeyError:
+        return None
+
+
+def get_details(reading: dict, detail: str) -> str | int | None:
+    """
+    Fetches specific details by searching through the keys
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    if not isinstance(detail, str):
+        raise TypeError('botanist_detail must be of type str')
+    return reading.get(detail)
+
+
+def botanist_details(reading: list[dict]) -> dict:
+    """
+    Returns a dataframe of botanist details
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
```

Comment:
This will execute every time. It should be:
```suggestion
    if not isinstance(reading,list):
        raise TypeError('entries into a DataFrame must be of type list')
```
But consider adding another test to ensure all values in reading are of type dict. For instance:
```
    if not all(isinstance(value, dict) for value in reading):
        raise TypeError('entries into a DataFrame must be of type dict')
```